### PR TITLE
Dynamic Chain Test Language Package With YAML Parser

### LINKED ISSUE #420: Test Definition Format: (Fork Choice Immediate Message Driven GHOST)
The test format for Casper fork choice rule is up for discussion [here](https://notes.ethereum.org/6ozACHyKR9i86vfkt086Ow#)
As a team, we should decide if we like the format and come up with a comprehensive design doc. It's also important we study what other client teams are doing for synchrony. (Ex: https://github.com/ethereum/beacon_chain/issues/58)
This issue is associated with #255 

More on fork choice rule here: https://ethresear.ch/t/immediate-message-driven-ghost-as-ffg-fork-choice-rule/2561

### PR DESCRIPTION
**Pull request by rauljordan**
Resolves #420

---

# Description

This PR incorporates the YAML format for fork-choice and chain tests as outlined in the Ethereum 2.0 spec and the markdown file [here](https://notes.ethereum.org/s/r11GVSBuQ) by @djrtwo. This PR does not yet implement the simulated backend that will act as the runner for the tests but simply includes a parser for supplied yaml files to extract test case parameters.

# Requirements for Merge
- [x] Basic Refactoring to Allow for Simulated Backend
- [x] Basic Simulated Backend Definition
- [x] Refactor Config to Allow for Custom Parameters for Simulated Runs
- [x] YAML Parser From File
- [x] Include a Basic Chaintest File
- [x] README on Using the Chaintest Runner

# Important

This PR does not actually run the full tests but instead just sets up the necessary framework and testing format required to do so. The test implementation will be part of #718.

### CODE REVIEW

File: beacon-chain/chaintest/main.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	prefixed "github.com/x-cray/logrus-prefixed-formatter"
+
+	"github.com/go-yaml/yaml"
+	"github.com/prysmaticlabs/prysm/beacon-chain/chaintest/backend"
+)
+
+func main() {
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/main.go]: Why is this a main binary rather than a go test?

File: beacon-chain/casper/sharding_test.go [hunk 2]
```diff
@@ -232,13 +232,13 @@ func TestValidatorsBySlotShardSmallValidatorSet(t *testing.T) {
 	for i := 0; i < len(shardAndCommitteeArray); i++ {
 		shardAndCommittees := shardAndCommitteeArray[i].ArrayShardAndCommittee
 		if len(shardAndCommittees) != 1 {
-			t.Fatalf("Expected %d committee per slot: got %d", params.GetConfig().MinCommiteeSize, 1)
+			t.Fatalf("Expected %d committee per slot: got %d", params.GetConfig().MinCommitteeSize, 1)
```
Comment (rauljordan) [inline beacon-chain/casper/sharding_test.go]: There were a few typos with Committee that needed to be fixed

File: .codecov.yml [hunk 3]
```diff
@@ -8,7 +8,10 @@ coverage:
   range: "70...100"
 
   status:
-    project: yes
+    project: 
+      default:
+        target: auto
+        threshold: 1%
```
Comment (prestonvanloon) [inline .codecov.yml]: Can we include this in a different PR?

Comment (prestonvanloon) [reply .codecov.yml]: nevermind, it looks like its already in master!

File: beacon-chain/chaintest/backend/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+go_library(
+    name = "go_default_library",
+    srcs = [
+        "setup_db.go",
+        "simulated_backend.go",
+        "test_format.go",
+    ],
+    importpath = "github.com/prysmaticlabs/prysm/beacon-chain/chaintest/backend",
+    visibility = ["//beacon-chain:__subpackages__"],
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/BUILD.bazel]: Please limit to `//beacon-chain/chaintest:__subpackages__`

File: beacon-chain/chaintest/backend/setup_db.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+	"path"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/prysmaticlabs/prysm/beacon-chain/db"
+)
+
+// setupDB instantiates and returns a simulated backend BeaconDB instance.
+func setupDB() (*db.BeaconDB, error) {
+	path := path.Join(os.TempDir(), "/simulateddb")
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/setup_db.go]: Will this cause issues if you are running tests in parallel? 
Maybe a random named directory would be ideal.

File: beacon-chain/chaintest/backend/simulated_backend.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
+// Package backend contains utilities for simulating an entire
+// ETH 2.0 beacon chain in-memory for e2e tests and benchmarking
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/simulated_backend.go]: Is this true? The database writes to disk so its not fully in-memory. Was that the intention?

File: beacon-chain/chaintest/backend/simulated_backend.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	for i := uint64(0); i < testCase.Config.ValidatorCount; i++ {
+		validators[i] = &pb.ValidatorRecord{
+			Status:            uint64(params.Active),
+			Balance:           currentConfig.DepositSize * currentConfig.Gwei,
+			WithdrawalAddress: []byte{},
+			Pubkey:            []byte{},
+			RandaoCommitment:  randaoReveal[:],
+		}
+	}
+	// TODO(#718): Next step is to update and save the blocks specified
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/simulated_backend.go]: Please add to your PR description that this PR does not actually run the tests and that the test implementation is part of #718

**Comment by rauljordan** [comment]
PTAL @prestonvanloon

File: beacon-chain/chaintest/backend/setup_db.go [hunk 8]
```diff
@@ -25,7 +27,7 @@ func teardownDB(db *db.BeaconDB) {
 	if err := db.Close(); err != nil {
 		log.Fatalf("failed to close database: %v", err)
 	}
-	path := path.Join(os.TempDir(), "/simulateddb")
+	path := path.Join(os.TempDir(), fmt.Sprintf("/%d", rand.Int()))
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/setup_db.go]: This won't clean up the same random number will it?

I think you need to keep track of this number somehow, maybe this path could be `db.DatabasePath`

File: beacon-chain/chaintest/backend/setup_db.go [hunk 9]
```diff
@@ -7,12 +7,14 @@ import (
...
 
+	"math/rand"
+
 	"github.com/prysmaticlabs/prysm/beacon-chain/db"
 )
 
 // setupDB instantiates and returns a simulated backend BeaconDB instance.
 func setupDB() (*db.BeaconDB, error) {
-	path := path.Join(os.TempDir(), "/simulateddb")
+	path := path.Join(os.TempDir(), fmt.Sprintf("/%d", rand.Int()))
```
Comment (prestonvanloon) [inline beacon-chain/chaintest/backend/setup_db.go]: Please prefix with something human readable: `prysmsimulateddb-%d` or similar

File: beacon-chain/chaintest/backend/setup_db.go (hunk 8 above)
Comment (terenc3t) [reply beacon-chain/chaintest/backend/setup_db.go]: +1

File: beacon-chain/chaintest/backend/setup_db.go (hunk 9 above)
Comment (nisdas) [reply beacon-chain/chaintest/backend/setup_db.go]: +1

File: beacon-chain/chaintest/backend/setup_db.go (hunk 8 above)
Comment (nisdas) [reply beacon-chain/chaintest/backend/setup_db.go]: Yeah agreed, maybe try to keep track of the number of simulated backend instances running and assign your path that way.

**Pull request merged**